### PR TITLE
#405 Fix "File name too long" error with deep paths in JUnit reports

### LINKED ISSUE #405: "File name too long" error with deep paths in JUnit reports
When using the `htmlSanityCheck` Gradle plugin within a sub-project, I encounter a "File name too long" error during the generation of JUnit reports. 

This issue arises because the generated report filenames incorporate both the full path to the Gradle sub-project and the deeply nested folder structure of the files being checked, exceeding some filename length limit.

```
Caused by: java.lang.RuntimeException: java.io.FileNotFoundException: /home/xxxxx/xxx/xxxxxxxx/xxxxxxxxxx/xxxxxxxxxxxx/xxxxxxxxxxxxxxxxxxxxx/documentation/build/test-results/htmlchecks/TEST-unit-html-_xxxx_xxxxx_xxx_xxxxxxxx_xxxxxxxxxx_xxxxxxxxxxxx_xxxxxxxxxxxxxxxxxxxxx_documentation_build_test-results_htmlchecks_xxxxxxxxxxxxxxx_xx_xxxxx_xxxxxxxx_xxx_xxxx_xxxxxxxxx_xx_xxxx_xx_xx_xxxxxxxxxxxxx_xx_xxxxxx_xxxxxxxxxxxxx_xx_xxxxxxxx_xxxxxxxxxx_xxxx.xml (File name too long)
        at org.aim42.htmlsanitycheck.report.JUnitXmlReporter.reportPageSummary(JUnitXmlReporter.java:95)
...
Caused by: java.io.FileNotFoundException: /home/xxxxx/xxx/xxxxxxxx/xxxxxxxxxx/xxxxxxxxxxxx/xxxxxxxxxxxxxxxxxxxxx/documentation/build/test-results/htmlchecks/TEST-unit-html-_xxxx_xxxxx_xxx_xxxxxxxx_xxxxxxxxxx_xxxxxxxxxxxx_xxxxxxxxxxxxxxxxxxxxx_documentation_build_test-results_htmlchecks_xxxxxxxxxxxxxxx_xx_xxxxx_xxxxxxxx_xxx_xxxx_xxxxxxxxx_xx_xxxx_xx_xx_xxxxxxxxxxxxx_xx_xxxxxx_xxxxxxxxxxxxx_xx_xxxxxxxx_xxxxxxxxxx_xxxx.xml (File name too long)
        at org.aim42.htmlsanitycheck.report.JUnitXmlReporter.reportPageSummary(JUnitXmlReporter.java:62)
```

### PR DESCRIPTION
**Pull request by ascheman**
## Summary

Fixes #405 - Resolves "File name too long" error when generating JUnit XML reports for HTML files in deeply nested directory structures.

The solution introduces a configurable output style for JUnit XML reports:
- **FLAT** (default): Maintains backward compatibility with existing behavior - all files in one directory with encoded paths
- **HIERARCHICAL**: New mode that mirrors source file directory structure, solving filename length issues

### Core Implementation
- Added `JunitOutputStyle` enum as nested class in `Configuration` (FLAT, HIERARCHICAL)
- Refactored `JUnitXmlReporter` to support both output modes with strategy pattern
- Implemented `getHierarchicalOutputFile()` with path normalization and security checks
- Updated `AllChecksRunner` to pass configuration to JUnitXmlReporter

### Security & Quality Improvements (from GitHub Copilot review)
- **Enhanced path traversal security**: Replaced string-based path validation with NIO Path API for robust protection against sophisticated attacks
- **Improved error diagnostics**: Added detailed context to directory creation failures including existence status and parent permissions
- **Comprehensive security testing**: Added tests for path traversal attacks and error message validation

### Plugin/CLI Updates
- **Gradle plugin**: Added `junitOutputStyle` input property
- **Maven plugin**: Added `junitOutputStyle` parameter
- **CLI**: Added `--junitOutputStyle` / `-o` command-line option

### Testing
- Added 7 comprehensive tests for HIERARCHICAL mode including edge cases
- Added 2 tests for FLAT mode and exception handling
- Added 5 tests for security (path traversal) and enhanced error messages
- Improved overall test coverage for JUnitXmlReporter
- All 384+ tests pass successfully

### Documentation
- Created comprehensive AsciiDoc documentation in `src/docs/development/_includes/issue-405.adoc`
- Includes problem description, solution details, configuration examples, and implementation notes

## Testing with JitPack

You can test this fix before it's released by using JitPack to build from this branch:

### Gradle
\`\`\`groovy
repositories {
    maven { url 'https://jitpack.io' }
}

dependencies {
    // For the Gradle plugin
    classpath 'com.github.aim42.htmlSanityCheck:htmlSanityCheck-gradle-plugin:bugfix~405-enable-junit-results-per-dir-SNAPSHOT'
}

htmlSanityCheck {
    junitOutputStyle = org.aim42.htmlsanitycheck.Configuration.JunitOutputStyle.HIERARCHICAL
}
\`\`\`

### Maven
\`\`\`xml
<repositories>
    <repository>
        <id>jitpack.io</id>
        <url>https://jitpack.io</url>
    </repository>
</repositories>

<dependency>
    <groupId>com.github.aim42.htmlSanityCheck</groupId>
    <artifactId>htmlSanityCheck-maven-plugin</artifactId>
    <version>bugfix~405-enable-junit-results-per-dir-SNAPSHOT</version>
</dependency>

<configuration>
    <junitOutputStyle>HIERARCHICAL</junitOutputStyle>
</configuration>
\`\`\`

**Note**: JitPack replaces `/` with `~` in branch names, so `bugfix/405-enable-junit-results-per-dir` becomes `bugfix~405-enable-junit-results-per-dir-SNAPSHOT`

## Backward Compatibility

✅ **Fully backward compatible** - The default value is `FLAT`, maintaining existing behavior unless explicitly configured to use `HIERARCHICAL` mode.

## CI Status

All workflows passing:
- ✅ Build and Test
- ✅ HTML Sanity Check Matrix Test (Java 8, 11, 17, 21 on Ubuntu, macOS, Windows)
- ✅ GitHub Pages
- ✅ SonarCloud Quality Gate

## Related Commits

- 7116615 - Initial fix with hierarchical structure
- d076086 - Made feature configurable for backward compatibility  
- 3fad99e - Improved test coverage for SonarCloud
- bf8ce30 - Enhanced path security and error handling (Copilot review)

### CODE REVIEW

**Comment by ascheman** [comment]
@strido This PR addresses your issue #405. Could you please test the solution using JitPack (instructions in the PR description) to verify it resolves your "File name too long" error? Your feedback would be greatly appreciated!

**Comment by ascheman** [comment]
@copilot Please review the latest changes addressing your previous security and error handling suggestions.

File: htmlSanityCheck-core/src/main/java/org/aim42/htmlsanitycheck/report/JUnitXmlReporter.java [hunk 1]
```diff
@@ -96,6 +114,73 @@ protected void reportPageSummary(SinglePageResults singlePageResults) {
...
+
+        // Ensure the directory exists
+        if (!testOutputDir.exists() && !testOutputDir.mkdirs()) {
+            StringBuilder errorMsg = new StringBuilder("Cannot create directory: ")
+                .append(testOutputDir.getAbsolutePath());
+            errorMsg.append(" (exists: ").append(testOutputDir.exists())
+                    .append(", parent canWrite: ")
+                    .append(testOutputDir.getParentFile() != null ? testOutputDir.getParentFile().canWrite() : "unknown")
+                    .append(")");
+            throw new RuntimeException(errorMsg.toString()); //NOSONAR(S112)
```
[the remark this answers is not in the document]
Comment (ascheman) [reply htmlSanityCheck-core/src/main/java/org/aim42/htmlsanitycheck/report/JUnitXmlReporter.java:176]: Fine by now

File: htmlSanityCheck-core/src/main/java/org/aim42/htmlsanitycheck/report/JUnitXmlReporter.java [hunk 2]
```diff
@@ -96,6 +113,63 @@ protected void reportPageSummary(SinglePageResults singlePageResults) {
...
+        File testOutputDir;
+        if (parentDir != null) {
+            // Normalize the path to handle relative references like ".."
+            // This ensures we stay within the outputPath and don't try to escape it
+            try {
+                File tempPath = new File(outputPath, parentDir.getPath());
+                testOutputDir = tempPath.getCanonicalFile();
+
+                // Verify the canonical path is still under outputPath
+                if (!testOutputDir.getAbsolutePath().startsWith(outputPath.getCanonicalPath())) {
```
[the remark this answers is not in the document]
Comment (ascheman) [reply htmlSanityCheck-core/src/main/java/org/aim42/htmlsanitycheck/report/JUnitXmlReporter.java]: Resolved by @claude (based on this @copilot proposal).

**Comment by strido** [comment]
@ascheman I have tried the fix and its working with the new `HIERARCHICAL` junit output style on our project!

I tried the instructions in the description but haven't been able to get it to work (it seems like the build on jitpack fails?).
So I have built the branch and tested this locally, I guess that doesn't matter.

Thank you for your work!

**Comment by ascheman** [comment]
> @ascheman I have tried the fix and its working with the new `HIERARCHICAL` junit output style on our project!
> 
> I tried the instructions in the description but haven't been able to get it to work (it seems like the build on jitpack fails?). So I have built the branch and tested this locally, I guess that doesn't matter.
> 
> Thank you for your work!

Thanks for trying this, @strido. I experienced similar problems on another branch/PR (cf. #429 / #431) with Jitpack meanwhile.
Sometimes, Jitpack can be a nightmare. So I raised a [new issue](https://github.com/aim42/htmlSanityCheck/issues/432) to make the package (branches) available on GitHub packages. I hope to have a first publication available in the next days, stay tuned.

Sorry for the inconvenience.
